api/nakadi-event-bus-api.yaml

```diff
@@ -0,0 +1,1232 @@
+    later moment this will be configurable and not be inherent part of this API).
+
+    * Topic API (low level).
+
+    Other aspects of the Event Bus are at this moment to be defined and otherwise specificied, not included
+    in this version of this specification.
+
+    Notable omissions here are:
+    * The contract between Nakadi and clients in the context of the high-level API (i.e. the process for
+    clients to establish subscriptions).
+
+    * Submission of events to topics in batched manner.
+
+    * Security scopes (OAuth) for the different operations
+
+    * Validation and Enrichment Strategies available in Nakadi are out of this document an can be consulted
+    in an external document (TBD link to document).
+
+
+  version: '0.5'
+  contact:
+    name: Team Aruha @ Zalando
+    email: team-aruha+nakadi-maintainers@zalando.de
+schemes:
+  - https
+consumes:
+  - application/json
+produces:
+  - application/json
+securityDefinitions:
+  oauth2:
+    type: oauth2
+    flow: implicit
+    authorizationUrl: 'https://auth.example.com/oauth2/tokeninfo'
+
+paths:
+  /metrics:
+    get:
+      tags:
+        - monitoring
+      summary: Get monitoring metrics
+      responses:
+        '200':
+          description: Ok
+          schema:
+            $ref: '#/definitions/Metrics'
+        '401':
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+        '500':
+          description: Server error
+          schema:
+            $ref: '#/definitions/Problem'
+        '503':
+          description: Not available
+          schema:
+            $ref: '#/definitions/Problem'
+  /event-types:
+    get:
+      tags:
+        - schema-registry-api
+      description: Returns a list of all registered `EventType`s
+      parameters:
+        - name: X-Flow-Id
+          in: header
+          description: |
+            The flow id of the request, which is written into the logs and passed to called services. Helpful
+            for operational troubleshooting and log analysis.
+          type: string
+          format: flow-id
+      responses:
+        '200':
+          description: Ok
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/EventType'
+        '401':
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+        '500':
+          description: Server error
+          schema:
+            $ref: '#/definitions/Problem'
+
+    post:
+      tags:
+        - schema-registry-api
+      description: |
+        Creates a new `EventType`.
+
+        **Implementation note:** The creation of an EventType implicitly creates the structures in the
+        backing messaging implementation needed for the reception and persistency of the Events. Considering
+        that at this time only Kafka is used, this corresponds to the creation of a Topic. If so desired,
```

Comment:
@valgog Thanks for reply! Let me refer to your answers in order:

> That also means, that these event-types should have the same ordering key (partitioning key). 

Yes, that's true and I don't see that being an issue or a complexity - producers and consumers should understand this semantics and should adhere to it. Was that considered to be a problem?

> applications should be made responsible of synchronisation of the the flow in such a unified topic and this is much more difficult to achieve then for one event-type. Especially if entities are produced by separate applications (that do not have the means of strong synchronisation usually)

What do you mean by "synchronisation" of the flow? When I think about Nakadi or any other generic "event bus" / dumping ground for time-ordered data, I'm assuming that you just want to push the events there in order they appear in your system. I believe that in a microservice architecture it should be consumer's role to deal with issues like this **if they need it** (producers are dumb, whatever happens, they "record" it as an event) and by consumer here I mean the client application - the one that handles the data (e.g. the app that's processing the time-ordered data) - not the queue consumer (not e.g. Kafka consumer).

Still, I might be missing something and the use-cases I'm thinking about might be somehow limited, so could you please give me an example which you thought about when discussing that problem? I mean - when is it unacceptable for different, non-synchronised services to write events out of order, while it's still OK for them to possibly write them with out-of-order timestamps to their local databases for example? Let's say you want to replay all the events to the queue because of an error in the processing that caused data corruption - if your sources of truth (local DBs) are possibly out of sync in terms of ordering, you still have a problem that you can't fix. 

> we definitely need a correlation engine (probably on the consumer side

Could you give me more details on how you think it should work? By consumer side do you mean something on Nakadi side that happens when client (consumer app) connects for consumption? Would that be in memory? How would you guarantee the ordering between low and high velocity topics then? Imagine an analytics engine for e.g. online store - you have topics with events like PageView (thousands / millions of events per minute) and ProductPurchased (tens of events per minute). How would you define the consumption time window / batch to make sure that you can properly order e.g. 2M PageView events produced over a minute and 10 ProductPurchased events produced in the same time? What if the "large" topic consumption can't catch up? You're reading ProductPurchase events as soon as they appear in the topic, but you're still little behind with PageViewed ones - do you stop consumption of the ProductPurchased and wait for PageViews to catch up? I think you're still not guaranteed to get all the messages in a given time window - even if the consumption time window is much larger than the "time window" used for returning data to the client app to allow you for accepting messages that are "late", it's still not protecting you from all the possible cases. I'm also a bit worried about consumption performance and resource requirements in that case (re-ordering time, memory, consumption delay) as an extra post-processing step is required and it's definitely non-trivial. 

Sorry for the huge number of questions, I hope I haven't lost / bored you somewhere on the way here ( ;-) ), but I'm quite concerned with this approach - I probably need some clarification on what you have in mind because at the moment (as I make a lot of assumptions :-) ) it seems to me like we're picking the right tool to do the job (Kafka), we remove from it what it's really good at (time-ordering, fast consumption) and then add another layer that's doing the same thing again but in less performant way and still allowing for out-of-order messages ;-) 

If I misunderstood the idea of correlation engine, please let me know how you see it as that's probably the most confusing part for me :-)